File: typescript/src/mcp/transport.ts
/**
 * Centralized MCP transport creation for backends.
 *
 * This module provides the core logic for creating MCP transports
 * based on backend type. Used by backends to implement getMCPTransport().
 */

import { StdioClientTransport } from '@modelcontextprotocol/sdk/client/stdio.js'
import type { StreamableHTTPClientTransport } from '@modelcontextprotocol/sdk/client/streamableHttp.js'
import type { Backend } from '../backends/types.js'
import { BackendType } from '../backends/types.js'
import { ERROR_CODES } from '../constants.js'
import { BackendError } from '../types.js'
import {
  getProperty,
  getRootBackend,
  hasRemoteConfig,
  isFileBasedBackend,
} from '../typing.js'
import { createAgentBeMCPTransport } from './client.js'
import { createLocalMCPTransportOptions, createMemoryMCPTransportOptions } from './local-transport.js'

/**
 * Union type for MCP transports
 */
export type MCPTransport = StdioClientTransport | StreamableHTTPClientTransport

/**
 * Create an MCP transport for a backend.
 *
 * This is the centralized implementation used by all backends.
 * Transport type depends on backend type:
 * - LocalFilesystemBackend → StdioClientTransport (spawns subprocess)
 * - RemoteFilesystemBackend → StreamableHTTPClientTransport (HTTP)
 * - MemoryBackend → StdioClientTransport (spawns subprocess)
 *
 * @param backend - The backend to create a transport for
 * @param scopePath - Optional scope path to override rootDir
 * @returns MCP transport configured for the backend type
 */
export async function createBackendMCPTransport(
  backend: Backend,
  scopePath?: string
): Promise<MCPTransport> {
  const rootBackend = getRootBackend(backend)
  const backendType = rootBackend.type

  switch (backendType) {
    case BackendType.LOCAL_FILESYSTEM:
      return createLocalTransport(backend, rootBackend, scopePath)

    case BackendType.REMOTE_FILESYSTEM:
      return createRemoteTransport(backend, rootBackend, scopePath)

    case BackendType.MEMORY:
      return createMemoryTransport(backend, scopePath)

    default:
      throw new BackendError(
        `Unsupported backend type: ${backendType}. getMCPTransport supports LOCAL_FILESYSTEM, REMOTE_FILESYSTEM, and MEMORY backends.`,
        ERROR_CODES.INVALID_CONFIGURATION,
        'unsupported-backend-type'
      )
  }
}

/**
 * Create Stdio transport for local filesystem backend
 */
async function createLocalTransport(
  backend: Backend,
  rootBackend: Backend,
  scopePath?: string
): Promise<StdioClientTransport> {
  const defaultRootDir = isFileBasedBackend(backend) ? backend.rootDir : '/'
  const rootDir = scopePath || defaultRootDir
  const isolation = getProperty<string>(rootBackend, 'isolation') ||
                    getProperty<string>(rootBackend, 'actualIsolation')
  const shell = getProperty<string>(rootBackend, 'shell')

  const options = createLocalMCPTransportOptions({
    rootDir,
    isolation: isolation as 'auto' | 'bwrap' | 'software' | 'none' | undefined,
    shell,
  })

  return new StdioClientTransport(options)
}

/**
 * Create HTTP transport for remote filesystem backend
 */
async function createRemoteTransport(
  backend: Backend,
  rootBackend: Backend,
  scopePath?: string
): Promise<StreamableHTTPClientTransport> {
  if (!hasRemoteConfig(rootBackend)) {
    throw new BackendError(
      'RemoteFilesystemBackend requires host to be configured. ' +
      'The MCP server must run on the remote host and be accessible via HTTP.',
      ERROR_CODES.INVALID_CONFIGURATION,
      'host'
    )
  }

  const { config } = rootBackend
  const mcpHost = config.mcpServerHostOverride || config.host
  const mcpPort = config.mcpPort || 3001
  const defaultRootDir = isFileBasedBackend(backend) ? backend.rootDir : '/'

  return createAgentBeMCPTransport({
    url: `http://${mcpHost}:${mcpPort}`,
    authToken: config.mcpAuth?.token || '',
    workspaceRoot: scopePath || defaultRootDir,
    userId: '',
    workspace: '',
  })
}

/**
 * Create Stdio transport for memory backend
 */
async function createMemoryTransport(
  backend: Backend,
  scopePath?: string
): Promise<StdioClientTransport> {
  const defaultRootDir = isFileBasedBackend(backend) ? backend.rootDir : '/'
  const rootDir = scopePath || defaultRootDir
  const options = createMemoryMCPTransportOptions({ rootDir })

  return new StdioClientTransport(options)
}

File: packages/agent-backend/python/agent_backend/mcp_integration/transport.py
"""MCP transport factory.

Creates the appropriate MCP transport based on backend type:
- LocalFilesystemBackend -> StdioServerTransport (spawns subprocess)
- RemoteFilesystemBackend -> StreamableHTTPClientTransport (HTTP)
- MemoryBackend -> StdioServerTransport (spawns subprocess)
"""

from __future__ import annotations

from typing import Any

from agent_backend.types import BackendType


async def create_backend_mcp_transport(
    backend: Any,
    scope_path: str | None = None,
) -> Any:
    """Create an MCP transport for a backend.

    Transport type depends on backend type:
    - local-filesystem -> stdio (spawns agent-backend daemon subprocess)
    - remote-filesystem -> HTTP
    - memory -> stdio (spawns agent-backend daemon subprocess)
    """
    from mcp.client.stdio import StdioServerParameters

    backend_type = backend.type

    if backend_type == BackendType.LOCAL_FILESYSTEM:
        root_dir = backend.root_dir
        effective_root = f"{root_dir}/{scope_path}" if scope_path else root_dir

        args = ["daemon", "--rootDir", effective_root, "--local-only"]

        isolation = getattr(backend, "_isolation", None)
        if isolation:
            args.extend(["--isolation", str(isolation)])

        shell = getattr(backend, "_shell", None)
        if shell:
            args.extend(["--shell", str(shell)])

        params = StdioServerParameters(command="agent-backend", args=args)
        return _StdioTransportWrapper(params)

    elif backend_type == BackendType.REMOTE_FILESYSTEM:
        config = backend.config
        mcp_host = config.mcp_server_host_override or config.host
        # Same port as the SSH-WS channel (see RemoteFilesystemBackend)
        mcp_port = config.port or config.mcp_port or 3001
        root_dir = backend.root_dir

        from agent_backend.backends.transports.daemon_endpoint import daemon_scheme
        from agent_backend.mcp_integration.client import create_http_transport

        scheme = daemon_scheme("http", mcp_port, config.secure)
        return create_http_transport(
            url=f"{scheme}://{mcp_host}:{mcp_port}",
            auth_token=config.auth_token or "",
            root_dir=root_dir,
            scope_path=scope_path,
            headers=config.headers,
        )

    elif backend_type == BackendType.MEMORY:
        root_dir = backend.root_dir
        effective_root = f"{root_dir}/{scope_path}" if scope_path else root_dir

        args = ["--backend", "memory", "--rootDir", effective_root]

        params = StdioServerParameters(command="agent-backend", args=args)
        return _StdioTransportWrapper(params)

    else:
        from agent_backend.types import BackendError, ErrorCode

        raise BackendError(
            f"Unsupported backend type: {backend_type}",
            ErrorCode.INVALID_CONFIGURATION,
            "unsupported-backend-type",
        )


class _StdioTransportWrapper:
    """Wraps StdioServerParameters for deferred connection."""

    def __init__(self, params: Any) -> None:
        self.params = params
        self._process: Any = None

    async def close(self) -> None:
        if self._process:
            try:
                self._process.terminate()
            except Exception:
                pass
